### DISCUSSION TITLE
Battery SoC for Nissan Leaf using MQTT

### DISCUSSION THREAD

**Discussion by camueller** [meatpiHQ/wican-fw#93] [Q&A]
Today I received my meatPi with firmware 2.9.0. After I configured Wifi and MQTT the MQTT Explorer displays the status message of the meatPi: `{"status": "online"}`.

I would like to read the SoC of my Nissan Leaf ZE1. According to your [documentation](https://github.com/meatpiHQ/wican-fw?tab=readme-ov-file#4-transmit-frames) I used the topic `wican/5432048f421d/can/tx` to publish the message
```
{"bus":"0","type":"tx","frame":[{"id":2015,"dlc":8,"rtr":false,"extd":false,"data":[2,1,91,170,170,170,170,170]}]}
```

However no response is displayed in MQTT Explorer. Any ideas what I'm doing wrong?

I left default CAN settings being:
- CAN bitrate: 500K (should be fine for Nissan Leaf according to the information I found)
- CAN mode: normal
- Port type: TCP
- Port: 3333
- Protocol: slcan

**Comment by camueller** [comment]
My bad -  car was not switched on :-(

Now I'm getting responses :-)

I found a [CAN documentation for the Nissan Leaf 2018](https://drive.google.com/file/d/1jH9cgm5v23qnqVnmZN3p4TvdaokWKPjM/view).
It states:

>The Li-ion battery controller accepts the multi-message query only.
> Query:
> 0x79B 02 21 01 00 00 00 00 00
> 0x79B 30 00 00 00 00 00 00 00

How can I send a multi message query using MQTT?

**Reply by meatpiHQ** [reply]
Can you please post the carscanner log when your reading the SoC. 

Have a look at this guide if you haven't yet.

https://github.com/meatpiHQ/wican-fw/wiki/EV-Battery-SoC-in-Home-Assistant-%E2%80%90-Example

**Comment by camueller** [comment]
In order to use the Car Scanner app I have to enable BLE in meatPi which somehow does not seem to work:
After I change "BLE status" from "disable" to "enable" the "BLE status" field is disabled (displayed in grey) immedialy!?
After I pressed "Submit changes" the meatPi rebooted but the "BLE status" still shows "disabled" and I don't find the meatPi Blutooth device using my Android phone. Unfortunatly the "Status" does not display anything regarding BLE status (only Wifi and CAN).

**Reply by meatpiHQ** [reply]
@camueller why do you need BLE? You can use wifi.

Please post a screenshot of the device configuration page.

**Reply by camueller** [reply]
![image](https://github.com/meatpiHQ/wican-fw/assets/4489628/0d552a04-01ed-4dd8-b084-e678764a1c2e)

**Comment by meatpiHQ** [comment]
Change the mode to AP, disable mqtt and set the protocol to elm327.  The enable BLE. 

Have another look at the documentation .

**Comment by camueller** [comment]
Car Scanner screens are quite full of options an links. That's why I missed "Wi-Fi" initially.
I disabled mqtt and set the protocol to elm327 as you suggested.
In "Car scanner" I configured:

Wi-Fi
OBD2 adapter IP address: 192.168.69.124 (which is also shown by meatPi on the status page as "Station IP")
OBD2 adapter port: 35000 / 3333 (last one is displayed on meatPi status page as TCP/UDP Port) - I tried both ports

Car Scanner always display "Connection problems - Car Scanner wasn't able to establish connection with ELM327)"

**Comment by camueller** [comment]
After a reboot Car Scanner was able to connect and cycled throught a 17 step handshake. However, Car Scanner finally displayed:

"Connection problems
Car Scanner has successfully connected to ELM327, but ELM327 wasn't able to establish connection with your car."

**Comment by camueller** [comment]
ELM and ECU connect now and Car Scanner displays data in Dashboard!
However I cannot find the SoC anywhere. I have values like HV Battery Current, HV Battery Power, Voltage of every cell but no SoC. Any suggestion where to look?

**Reply by meatpiHQ** [reply]
Maybe car scanner does not know the SoC PID. I have an idea, I will send you something to test soon.

**Comment by camueller** [comment]
According to the Nissan Leaf CAN document I linked in my second post in this thread the Li-Ion Battery Controller has to be queried with PID 0x079B = 1947. The list of sensors in Car Scanner does not contain an entry with this ID. Perhaps a sensor has to be added for it.

**Reply by meatpiHQ** [reply]
Yes carscanner allows for custom PID, you can try and add that.

But what i think you should be looking at is this : https://github.com/meatpiHQ/wican-fw/discussions/86#discussion-6190245

and change the frames to match your car 

<img width="1171" alt="image" src="https://github.com/meatpiHQ/wican-fw/assets/94690098/f163ffca-3812-42f4-9721-f53851104fb9">

**Reply by meatpiHQ** [reply]
After looking at the SoC equation, it seems the only way to do this is using NodRED, because the SoC is in 2 frames.

what you need to do in NodRED is send these 2 frames, with 200ms delay between them. This is equivalent to the query they are doing in the document.
{"bus":"0","type":"tx","ts":4406,"frame":[{"id":1947,"dlc":8,"rtr":false,"extd":false,"data":[2,33,1,0,0,0,0,0]}]}
{"bus":"0","type":"tx","ts":4428,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,0,0,0,0,0]}]}

<img width="698" alt="image" src="https://github.com/meatpiHQ/wican-fw/assets/94690098/7d57803f-34e7-48bc-acb8-c054c0d7df37">

**Comment by camueller** [comment]
NodeRED I know well and use it myself but for SOC retrieval I want to use Linux shell scripts. For WiCAN I use `mosquitto_pub` and `mosquitto_sub` to interact with the MQTT server. Between two calls to `mosquitto_pub` in order to send both parts of the query I wait 100ms. From the MQTT server I receive 8 MQTT messages from which I can extract the SOC successfully.

Full shell scripts are available here:
https://github.com/camueller/SmartApplianceEnabler/raw/master/run/soc/wican/wican-status.sh
https://github.com/camueller/SmartApplianceEnabler/raw/master/run/soc/wican/wican-soc.sh

Thank you very much for providing a great product!

**Reply by Mathopenevse** [reply]
hi, if I understand, I need to send two data Json on topic "wican/id/can/tx" with about 2ms?
i have send 
`{"bus":"0","type":"tx","frame":{"id":1947,"dlc":8,"rtr":false,"extd":false,"data":[2,33,1,0,0,0,0,0]}}`
and
`{"bus":"0","type":"tx","frame":{"id":1947,"dlc":8,"rtr":false,"extd":false,"data":[48,0,0,0,0,0,0,0]}}`

but it doesn't work for me and I receive no data on "wican/id/can/rx" maybe I have missed something, if you could help me? thx

**Reply by camueller** [reply]
If you receive no message on "wican/id/can/rx" then most likely the car is not powered on and therefore the CAN bus is not operational. Or the WiCAN adapter is not connected to your Wifi.

**Reply by Mathopenevse** [reply]
thanks for your answer, I receive data now but after integrated "ts" 
![307762889-ae43b429-940c-4c95-bb19-8ab74f322458](https://github.com/meatpiHQ/wican-fw/assets/130982702/275502af-c909-421f-97e4-d5b4efab1f6d)
it's seems I should receive 8 trames but just only one in my case...

**Reply by camueller** [reply]
Most likely the data you want (probably SOC) is not contained in the first message. You should receive 8 messages:
```
Feb 27 14:52:10 raspi2 wican-soc.sh[27329]: Message received:
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6805,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[16,53,97,1,0,0,3,232]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6917,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[33,2,175,0,0,0,30,0]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6925,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[34,0,0,0,0,0,48,212]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6937,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[35,157,49,47,126,3,165,0]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6945,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[36,1,112,0,36,0,0,15]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6957,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[37,27,198,0,15,180,27,128]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6965,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[38,0,5,0,0,0,30,0]}]}
Feb 27 14:52:10 raspi2 wican-soc.sh[9865]: {"bus":"0","type":"rx","ts":6977,"frame":[{"id":1979,"dlc":8,"rtr":false,"extd":false,"data":[39,0,0,170,1,174,255,255]}]}
```
You may simply run the shell scripts linked above. They run on any Linux device ...

**Reply by Mathopenevse** [reply]
thanks! in fact MQTT-explorer show just the last frame on can/rx but i have receive 8 frames in the history! now i have to filter the different answer with the filter !
